File: Boids_2_c_raylib_myedit/src/subgen.c
#include "subgen.h"

#define MAX 1000000000
int state[55];
int si = 0;
int sj = 0;
int i = 0;
int j = 0;
int x = 0;
int p2 = 1;

void Subgen_Init(int seed)
{
	state[0] = seed % MAX;
	for (i = 1, j = 21; i < 55; i++, j += 21)
    {
		if (j >= 55) j -= 55;
		state[j] = p2;
        p2 = seed - p2;
		if (p2 < 0) p2 += MAX;
		seed = state[j];
	}
	si = 0;
	sj = 24;

	for (i = 0; i < 165; i++) Subgen_Next();
}
int Subgen_Next()
{
	if (si == sj) Subgen_Init(0);
	if (!si--) si = 54;
	if (!sj--) sj = 54;
    x = state[si] - state[sj];
	if (x < 0) x += MAX;
	return state[si] = x;
}
float Subgen_FractionUnsigned()
{
    return (float)Subgen_Next() / (float)MAX;
}
float Subgen_FractionSigned()
{
    return Subgen_FractionUnsigned() * 2 - 1;
}
float Subgen_Range(float min, float max)
{
    return min + ((max - min) * Subgen_FractionUnsigned());
}

File: Boids_wpf_3/src/SubGen.cs
public static class Subgen
{
    public static int MAX = 1000000000;
    private static int[] state = new int[55];
    private static int pos;

    private static int mod(int n)
    {
        return ((n % MAX) + MAX) % MAX;
    }
    public static void Init(int seed)
    {
        int[] temp = new int[55];
        temp[0] = mod(seed);
        temp[1] = 1;
        for(int i = 2; i < 55; ++i)
            temp[i] = mod(temp[i - 2] - temp[i - 1]);

        for(int i = 0; i < 55; ++i)
            state[i] = temp[(34 * (i + 1)) % 55];

        pos = 54;
        for(int i = 55; i < 220; ++i)
            Next();
    }
    public static int Next()
    {
        int temp = mod(state[(pos + 1) % 55] - state[(pos + 32) % 55]);
        pos = (pos + 1) % 55;
        state[pos] = temp;
        return temp;
    }
    public static float FractionUnsigned()
    {
        var next = Next();
        // System.Console.WriteLine(next);
        return next / (float)MAX;
    }
    public static float FractionSigned()
    {
        float result = FractionUnsigned() * 2 - 1;
        // Helper.PrintFloatHex(result);
        // System.Console.WriteLine();
        return result;
    }
    public static float Range(float min, float max)
    {
        float result = min + ((max - min) * FractionUnsigned());
        // Helper.PrintFloatHex(result);
        // System.Console.WriteLine();
        return result;
    }
}

/*
0.824359 = 824358983 / 1000000000.000000
0.925189 = 925189384 / 1000000000.000000
0.988325 = 988325069 / 1000000000.000000
0.030883 = 30882955 / 1000000000.000000
0.367453 = 367453453 / 1000000000.000000

0.824359 = 824358983 / 1E+09
0.9251894 = 925189384 / 1E+09
0.98832506 = 988325069 / 1E+09
0.030882956 = 30882955 / 1E+09
0.36745343 = 367453453 / 1E+09
*/